src/main/kotlin/parts/wisdom/deansher/arcraider/Arc.kt

```diff
@@ -0,0 +1,124 @@
+package parts.wisdom.deansher.arcraider
+
+import parts.wisdom.arcraider.ArcColor
+import parts.wisdom.arcraider.SerializedGrid
+import parts.wisdom.arcraider.ArcPair as SerializedPair
+import parts.wisdom.arcraider.Task as SerializedTask
+
+/**
+ * A grid with dimensions `width, height`, indexed by `[x, y]` where `x in 0 until width` goes left to right
+ * and `y in 0 until height` goes top to bottom.
+ */
+data class ArcGrid(
+    val width: Int,
+    val height: Int,
+    /** Stores `(x, y)` at `y * width + x` */
+    private val data: Array<ArcColor>
+) {
+    /**
+     * Construct an `ArcGrid` with all squares the same color.
+     */
+    constructor(width: Int, height: Int, color: ArcColor) :
+            this(width, height, Array(height * width) { color })
+
+    operator fun get(x: Int, y: Int): ArcColor = data[dataIndex(x, y)]
+
+    operator fun set(x: Int, y: Int, color: ArcColor) {
+        data[dataIndex(x, y)] = color
+    }
+
+    private fun dataIndex(x: Int, y: Int): Int {
+        require(x in 0 until width) { "invalid x=$y" }
+        require(y in 0 until height) { "invalid y=$x" }
+        return y * width + x
+    }
+
+    override fun toString(): String {
+        val rows = (0 until height).map { y ->
+            data.slice((y*width) until ((y+1)*width))
+        }
+        val rowStrings = rows.map { row ->
+            row.map { it.toChar() }.joinToString("")
+        }
+        return rowStrings.joinToString("\n")
+    }
+
+    override fun equals(other: Any?): Boolean {
```

Comment:
Is this function necessary? This is a data class and your just comparing the fields?